### DISCUSSION TITLE
How to modify the width or height of TableItem in pixels

### DISCUSSION THREAD

**Discussion by pcdlrzxx** [eclipse-platform/eclipse.platform.swt#1539] [Q&A] [answered]
I can simply modify the width and height of TableItem through the code below
```
table.addListener(SWT.MeasureItem, new Listener() {
   public void handleEvent(Event event) {
      event.width = 100;
      event.height = 20;
   }
});
```
but it only works in points, and I just want to know how to make it work in pixels.

As for Table, although it also can't do this by "Public" method, but can be achieved by using reflection.
However the "Event" class doesn't provide similar methods.

Can anyone help me ,thanks very much!

**Comment by elsazac** [comment]
I believe you can't directly set sizes in pixels in SWT. By default, it uses points, which adjust based on the display's DPI. To get pixel sizes, you need to manually do may be, pixel conversions and apply some sort of scaling or so.. As far as I know, there's no API that automatically switches between points and pixels, but please do correct me if I'm mistaken.

**Reply by pcdlrzxx** [reply]
Thanks for your reply.
> I believe you can't directly set sizes in pixels in SWT. 

In fact, there's the "setBoundsInPixels" method to set the location and size in pixels in Table class or Control class, however it is not "public", but we still can invoke by using reflection. So in this way, we can directly set sizes in pixels for table.

The problem I'm facing is, for example, when I set the scaling to 200% and I need a tableitem with 200 width in pixel, "event.width = 100;" works, but if I need a tableitem with 199 width in pixel, "event.width = 100;" or "event.width = 99;" can't help.

**Reply by elsazac** [reply]
Yes, `setBoundsInPixels` is an option but it positions and sizes the entire Table, like a Table or any other widget. To control the size of TableItem, we need `SWT.MeasureItem`, which manages the width and height of each item. You can only control the size using `SWT.MeasureItem`, but the positioning is handled by the control itself. 

I understand your point that while using event.width = 100 works for some sizes, it may not give the exact pixel width needed for values like 199 pixels. I'm not sure how to address that, but it might require manually converting and applying the correct scaling.

**Reply by pcdlrzxx** [reply]
Perhaps I should learn more and deep on this.
And I'm wondering why doesn't SWT set those `xxxInPixels` method to `public`, if public, invoking can be done directly without using reflection.
And why there is no such method like `setBoundsInPixels` in `TableItem` class, I think it can bring convenience on custom drawing with such methods.

**Reply by laeubi** [reply]
@pcdlrzxx maybe you should share a runnable snippet to show the problem, if you report 100 as a height and then draw a box with height 50 it should fill half of the area, so why do you need "pixels" at all (what is more a technical thing), so for your example why should the box drawn in your example on 200% only fill a quarter of the area?

**Reply by pcdlrzxx** [reply]
@laeubi Maybe I didn't express it clearly. What I really want to say is that perhaps there is something strange in size conversion from points to pixels in SWT. I will show a demo in the following:
```
import java.lang.reflect.Method;

import org.eclipse.swt.SWT;
import org.eclipse.swt.events.MouseAdapter;
import org.eclipse.swt.events.MouseEvent;
import org.eclipse.swt.graphics.Rectangle;
import org.eclipse.swt.layout.RowLayout;
import org.eclipse.swt.widgets.*;

public class ScaleTest {
	
	private static Text text;
	
	private static Table table;
	
	public static void main(String[] args) {
		System.setProperty("swt.autoScale", "100");
//		System.setProperty("swt.autoScale", "192");
//		System.setProperty("swt.autoScale", "198");
		
		Display display = new Display();
		final Shell shell = new Shell(display);
		shell.setLayout(new RowLayout(SWT.VERTICAL));
		shell.setSize(400, 300);
		
		Button button = new Button(shell, SWT.PUSH);
		button.setText("bounds info");
		button.addMouseListener(new MouseAdapter() {
			@Override
			public void mouseUp(MouseEvent e) {
				Rectangle textBounds = text.getBounds();
				System.out.println("text bounds in points: x=" + textBounds.x + ", y=" + textBounds.y + ", w=" + textBounds.width + ", h=" + textBounds.height);
				Rectangle tableBounds = table.getBounds();
				System.out.println("table bounds in points: x=" + tableBounds.x + ", y=" + tableBounds.y + ", w=" + tableBounds.width + ", h=" + tableBounds.height);
				
				Rectangle textBoundsInPixels = null;
				Rectangle tableBoundsInPixels = null;
				try {
					Method[] methods = Control.class.getDeclaredMethods();
					for (Method method : methods) {
						if (method.getName().equals("getBoundsInPixels")) {
							boolean accessible = method.isAccessible();
							method.setAccessible(true);
							textBoundsInPixels = (Rectangle) method.invoke(text);
							tableBoundsInPixels = (Rectangle) method.invoke(table);
							method.setAccessible(accessible);
							break;
						}
					}
				} catch (Exception ex) {
					ex.printStackTrace();
				}
				System.out.println("text bounds in pixels: x=" + textBoundsInPixels.x + ", y=" + textBoundsInPixels.y + ", w=" + textBoundsInPixels.width + ", h=" + textBoundsInPixels.height);
				System.out.println("table bounds in pixels: x=" + tableBoundsInPixels.x + ", y=" + tableBoundsInPixels.y + ", w=" + tableBoundsInPixels.width + ", h=" + tableBoundsInPixels.height);
				if (textBoundsInPixels.y + textBoundsInPixels.height == tableBoundsInPixels.y) {
					System.out.println("the table fit the text perfectly on vertical direction.");
				} else {
					System.out.println("the table fit the text not perfectly on vertical direction!");
				}
			}
		});
		
		text = new Text(shell, SWT.BORDER);
		text.addListener(SWT.Paint, new Listener() {
			@Override
			public void handleEvent(Event event) {
				text.setBounds(100, 20, 200, 20);
			}
		});
		
		table = new Table(shell, SWT.BORDER);
		table.addListener(SWT.Paint, new Listener() {
			@Override
			public void handleEvent(Event event) {
				Rectangle r = text.getBounds();
				table.setBounds(r.x, r.y + r.height, r.width, 50);
			}
		});
		
		shell.open();
		
		while (!shell.isDisposed()) {
			if (!display.readAndDispatch())
				display.sleep();
		}
		display.dispose();
	}
	
}
```
When I set `swt.autoScale=100`, it performs perfectly. But when set `swt.autoScale=192` or `swt.autoScale=198`, it performs a bit Incongruously. However, if I replace `setBounds` to `setBoundsInPixels` on the table, I believe it will always perform perfectly at all scaling ratios.
This demo is just an analogy. The real problem I am facing now is that `TableItem`'s width only can modify by `event.width` which is in points, and I can't make its width consistent with the width of other components like `Text` at some scaling ratios in pixels (either one more pixel or one less pixel). This makes me very frustrated.

**Reply by sravanlakkimsetti** [reply]
@pcdlrzxx Not making `xxxInPixels` methods as `public` is a conscious decision. In Linux and Mac we always work with points. The translation to pixels is at OS/native layer which we are not privy to. Only in case of Windows, SWT layer handles this translation. SWT API should be agnostic to underlying OS/windowing system. For this reason we have made the api to work points only. 

We had to go though `xxxInPixels` approach as at the time of HiDpi development the Windows and Linux did not have good HiDpi support natively. By GTK 3.18(not sure) release the linux had good HiDpi support. We leveraged this and revamped Linux support to directly work with points only.

Using `swt.autoScale` you specify scaling factor as far as remember. In windows we support only two scaling factors 100% and 200%. When you provide a value up to 174 the scaling factor selected will be 100% and 175 and above the scaling factor selected is 200%.

**Reply by sravanlakkimsetti** [reply]
Also see https://eclipse.dev/eclipse/news/4.6/platform.php#swt-autoscale. By the way the tweaks mentioned here work with Windows only. 

The tweaks will not work with either Linux or Mac

**Reply by pcdlrzxx** [reply]
@sravanlakkimsetti Hello sravanlakkimsetti, thanks for your reply. I have read the `swt-autoscale-tweaks` part of the article, and I noticed some discrepancies. According to the article, there are two cases in `swt.autoScale=integer` and `swt.autoScale=<value>`. When `swt.autoScale=integer`, 100% and 200% will only be used. But when `swt.autoScale=<value>`, the set value will be used.
In fact, the value I set in program whatever 192 or 198 is effective, this can be proved by the output in program.

So in my understanding, because of the not good HiDpi support of Windows natively, problems are inevitable at some scaling factors on Windows. And we can only hope for improvements in Windows' HiDpi support, right?

**Reply by sravanlakkimsetti** [reply, marked as answer]
@pcdlrzxx 
> So in my understanding, because of the not good HiDpi support of Windows natively, problems are inevitable at some scaling factors on Windows. And we can only hope for improvements in Windows' HiDpi support, right?

Yes you right. We went with 25% increments initially and it created lots of problems with the sizes. Converting pixels to points using integer arithmetic is bound cause trouble. Thats why we had to completely remove 25% increments and then supported integer scale factors only. 

By the way Mac uses floating point arithmetic and Linux supports only integer scaling factors.

**Reply by pcdlrzxx** [reply]
@sravanlakkimsetti Fine, thanks for you answer! It seems that I don't need to dwell on this aspect for the time being. Hope the Windows' HiDpi support improvements come out as soon as possible.
